### PR TITLE
fix building of live-sample pages from url

### PR DESCRIPTION
**Pull request by escattone**
This PR fixes the building of live-sample pages by URL. Currently, when in functional mode and you visit http://localhost:5000/en-US/docs/Learn/CSS/CSS_layout/Introduction, you'll see empty live-sample `iframe`'s:

image

With this PR, you'll see this:

image

### CODE REVIEW

File: kumascript/src/live-sample.js [hunk 1]
```diff
@@ -118,30 +118,20 @@ class LiveSampleID {
 
 function getLiveSampleIDs(slug, source) {
   // Given a slug and its raw source HTML, parses the source and returns
```
Comment (escattone) [inline kumascript/src/live-sample.js:120]: The changes in this file aren't necessary for this fix. I just wanted to simplify this function knowing that we don't care about sample ID's to be extracted from other documents.

**Comment by escattone** [comment]
@Gregoor I really like what you did with the building of live sample pages, i.e. just build the live-samples for the current page and if the current page embeds live samples from other pages and is being locally previewed, build the live-samples extracted from the other pages as requested. Nice! It's simpler and cleaner. Thank you!

File: build/index.js [hunk 2]
```diff
@@ -228,13 +228,29 @@ async function buildDocumentFromURL(url) {
...
   const document = Document.findByURL(documentURL);
-  const liveSamplePage = kumascript.buildLiveSamplePage(
-    document.url,
-    document.metadata.title,
-    (await kumascript.render(document.url))[0],
-    { id: sampleID, createFlaw() {} }
-  );
-  return liveSamplePage.html;
+  if (!document) {
+    return `unable to find ${documentURL}`;
```
Comment (Gregoor) [inline build/index.js]: I think we should turn this into a 
```
throw new Error(`No document found for ${documentURL}`);
```
At which point we should also add some error handling / status sending into the server code.

File: build/index.js [hunk 3]
```diff
@@ -228,13 +228,29 @@ async function buildDocumentFromURL(url) {
...
+  }
+  // Convert the lower-case sampleID we extract from the incoming URL into
+  // the actual sampleID object with the properly-cased live-sample ID.
+  for (const sampleIDObject of kumascript.getLiveSampleIDs(
```
Comment (Gregoor) [inline build/index.js:236]: I'm wondering if we can keep the simplicity of the previous approach by changing the middleware @peterbe brought back in to also hold on to the `originalURL` [before lowercasing it](https://github.com/Gregoor/yari/blob/fc31de0f6ad7ead5bbf3a3573c4be3152903a1c1/server/middlewares.js#L19). Express allows you to attach arbitrary data to the request object, which is what a lot of middlewares do. We'd then have to make sure to pass it into this function. And it'd restrict the building to only work with correctly cased URLs, which yeah. I don't know if we want that. I guess it's a trade-off between caller simplicity vs implementation simplicity.

**Review by Gregoor — COMMENTED** [review]
Thank you for the vigilant review and unearthing that problem! I'm wondering whether it'd be worth exploring an alternative. like the one outlined in my 2nd comment.

File: build/index.js (hunk 3 above)
Comment (escattone) [reply build/index.js:236]: That's a good thing to explore. I would prefer to keep the simplicity of the previous approach (by adding `originalURL` to the request object as you said), but it seems wrong to depend upon the proper casing in URL's (as you mentioned as well).

File: build/index.js (hunk 2 above)
Comment (escattone) [reply build/index.js]: @Gregoor Yep, that's better! I'll do that.

File: build/index.js [hunk 4]
```diff
@@ -228,13 +228,29 @@ async function buildDocumentFromURL(url) {
...
+  )) {
+    if (sampleIDObject.id.toLowerCase() === sampleID) {
+      const liveSamplePage = kumascript.buildLiveSamplePage(
+        document.url,
+        document.metadata.title,
+        (await kumascript.render(document.url))[0],
+        sampleIDObject
+      );
+      if (liveSamplePage.flaw) {
+        return liveSamplePage.flaw.errorMessage;
```
Comment (Gregoor) [inline build/index.js]: also something we should `throw` instead of `return`ing it (and, optimally also handle in server)

**Review by Gregoor — APPROVED** [review]

File: build/index.js [hunk 5]
```diff
@@ -245,12 +245,12 @@ async function buildLiveSamplePageFromURL(url) {
         sampleIDObject
       );
       if (liveSamplePage.flaw) {
-        return liveSamplePage.flaw.errorMessage;
+        throw liveSamplePage.flaw;
```
Comment (Gregoor) [inline build/index.js]: it's best practice to only throw error objects (even though JS unfortunately allows you to throw _anything_), reasons being: catching code tends to expect an error object and only with an error object do we get a stack trace

**Comment by escattone** [comment]
@Gregoor Ok, I think I've got it now. At least I like it much better. Let's see what you think.

**Comment by Gregoor** [comment]
big fan! thanks for fixing it and taking the time to refine

**Pull request merged**